Add a plugin to enforce valid links to Bugzilla bugs

stevekuznetsov opened the pull request:
Often, GitHub Issues are not the only mechanism by which organizations
will track bugs against a product. Bugzilla is a common external bug
tracker; the new bugzilla plugin allows validation that the pull request
title references a bug that is valid as configured for the specific
branch.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @cjwagner @fejta @smarterclayton @wking 
/cc @bbguimaraes @petr-muller @hongkailiu @droslean

petr-muller commented on prow/bugzilla/types.go:
```diff
@@ -0,0 +1,140 @@
+	// IsOpen is true if this bug is open, false if it is closed.
+	IsOpen bool `json:"is_open,omitempty"`
+	// IsCreatorAccesible is if true, this bug can be accessed by the creator of the bug, even if they are not a member of the groups the bug is restricted to.
+	IsCreatorAccesible bool `json:"is_creator_accessible,omitempty"`
```
Accesible->Accessible

stevekuznetsov commented:
Tested it here: https://github.com/openshift/release/pull/3899

stevekuznetsov commented on prow/plugins/bugzilla/bugzilla.go:
```diff
@@ -0,0 +1,334 @@
+	// ensure label state is correct. Do not propagate errors
+	// as it is more important to report to the user than to
+	// fail early on a label check.
+	currentLabels, err := gc.GetIssueLabels(e.org, e.repo, e.number)
```
They're one and the same in the GitHub API

stevekuznetsov commented on prow/plugins/bugzilla/bugzilla.go:
```diff
@@ -0,0 +1,334 @@
+			errors = append(errors, fmt.Sprintf("expected the bug to target the %q release, but no target release was set", *options.TargetRelease))
+		} else if *options.TargetRelease != bug.TargetRelease[0] {
+			valid = false
+			errors = append(errors, fmt.Sprintf("expected the bug to target the %q release, but it targets %q instead", *options.TargetRelease, bug.TargetRelease[0]))
```
Yeah, unclear -- it's served to us via the API, but as a list...

stevekuznetsov replied:
Also not fragile given I check for the length above

stevekuznetsov commented on prow/plugins/bugzilla/bugzilla.go:
```diff
@@ -0,0 +1,334 @@
+			} else {
+				message = fmt.Sprintf("on the %q branch, ", branch)
+			}
+			message += "valid bugs must "
```
`bytes.Buffer()` would be more efficient technically but I'm not sure we really care? Happy to change it

petr-muller replied:
Not that familiar with Prow codebase yet, not sure what are your standards. Most teams I was working in (but that was Java and Python) were avoiding string concatenation in cycles and non-trivial cases (like, 5+ concatenations?), but I agree there's little performance hit in this specific case.

fejta commented:
How would you feel about making this an external plugin?

stevekuznetsov commented:
> How would you feel about making this an external plugin?

I considered it but I'm not really sure what the benefit there is. We haven't done the work to make it simple yet so it entails:

 - copy-pasting hundreds of lines of boilerplate
 - a separate deployment to run the plugin
 - more overhead on making sure the image gets published, etc

What do we gain by it? What makes a plugin external?

cjwagner commented on prow/plugins/config.go:
```diff
@@ -982,3 +983,135 @@ func (pluginConfig *ProjectConfig) GetOrgColumnMap(org string) map[string]string
+func ResolveBugzillaOptions(parent, child BugzillaBranchOptions) BugzillaBranchOptions {
+	output := BugzillaBranchOptions{}
+
+	// populate with the parent
```
nit: You can instead just use `output := parent`.

cjwagner commented on prow/plugins/config.go:
```diff
@@ -982,3 +983,135 @@ func (pluginConfig *ProjectConfig) GetOrgColumnMap(org string) map[string]string
+	TargetRelease *string `json:"target_release,omitempty"`
+}
+
+func (o BugzillaBranchOptions) matches(other BugzillaBranchOptions) bool {
```
nit: Should we just use `reflect.DeepEqual` so that this doesn't have to be maintained as new option fields are added?

stevekuznetsov replied:
I didn't want to have us running `reflect` code since it's really slow and you could end up doing this a lot with nested config

cjwagner commented on prow/plugins/config.go:
```diff
@@ -982,3 +983,135 @@ func (pluginConfig *ProjectConfig) GetOrgColumnMap(org string) map[string]string
+type Bugzilla struct {
+	// Default settings mapped by branch in any repo in any org.
+	// The `*` wildcard will apply to all branches.
+	Default map[string]BugzillaBranchOptions `json:"default"`
```
The implementation looks solid to me, just wondering how common use cases that benefit from this are since it adds complexity.

stevekuznetsov replied:
Yeah, since the field validation on the Bugzilla bugs is going to be on the target release, you'll want to make sure all `release-1.13` branch PRs target `1.13` version or similar, across an org. If we had a config like that for `branch-protector` you couldn't do that

stevekuznetsov commented on prow/bugzilla/client_test.go:
```diff
@@ -0,0 +1,98 @@
+		endpoint: url,
+		client: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
Honestly I just cargo culted this from the GitHub client tests

cjwagner replied:
If the public constructor is sufficient I think it would be preferable to use that. It doesn't really matter though.

stevekuznetsov commented on prow/flagutil/bugzilla.go:
```diff
@@ -0,0 +1,81 @@
+
+// AddFlags injects Bugzilla options into the given FlagSet.
+func (o *BugzillaOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.endpoint, "bugzilla-endpoint", "", "Bugzilla's API endpoint.")
```
I went the flag route since it seemed most like the other options we had for endpoints? Do you think one is better than the other?

cjwagner replied:
Config options have the two advantages over flags that I mentioned, but its up to you.

stevekuznetsov commented on prow/plugins/bugzilla/bugzilla.go:
```diff
@@ -0,0 +1,333 @@
+			} else {
+				message = fmt.Sprintf("on the %q branch, ", branch)
+			}
+			message += "valid bugs must "
```
Yeah, but in that case you're choosing to turn this on with a nonsense config anyway ;)

cjwagner replied:
If that config is considered nonsense we should fail validation during config load.

stevekuznetsov replied:
I updated the help comment to be more helpful instead

stevekuznetsov commented:
@cjwagner thanks for the review! Addressed your comments, added the features to remove labels when a title changes to no longer reference a bug, when someone asks for a refresh on a PR without a bug reference, and the no-op when a PR changes and the bug ID referenced does not

stevekuznetsov commented:
/test pull-test-infra-yamllint

cjwagner commented on prow/plugins/bugzilla/bugzilla.go:
```diff
@@ -0,0 +1,389 @@
+	prevMat := titleMatch.FindStringSubmatch(changes.Title.From)
+	if prevMat == nil {
+		// title did not previously reference a bug
+		return nil, nil
```
If the title did not previously reference a bug, but does now we need to handle the event.

stevekuznetsov replied:
eek, nice catch

stevekuznetsov commented:
@cjwagner updated

cjwagner commented:
/lgtm

Aisuko commented:
The prow is working on my own host almost one week pass, but I got an issue today.

```
{"component":"hook","file":"prow/flagutil/github.go:70","func":"k8s.io/test-infra/prow/flagutil.(*GitHubOptions).Validate","level":"warning","msg":"missing required flag: please set to --github-token-path=/etc/github/oauth before June 2020","severity":"warning","time" │
│ {"component":"hook","file":"prow/flagutil/github.go:78","func":"k8s.io/test-infra/prow/flagutil.(*GitHubOptions).Validate","level":"warning","msg":"It doesn't look like you are using ghproxy to cache API calls to GitHub! This has become a required component of Prow an │
│ {"component":"hook","file":"prow/flagutil/bugzilla.go:45","func":"k8s.io/test-infra/prow/flagutil.(*BugzillaOptions).Validate","level":"info","msg":"empty -bugzilla-endpoint, will not create Bugzilla client","severity":"info","time":"2020-07-02T07:18:15Z"}             │
│ {"client":"github","component":"hook","file":"prow/github/client.go:574","func":"k8s.io/test-infra/prow/github.(*client).log","level":"info","msg":"User()","severity":"info","time":"2020-07-02T07:18:15Z"}                                                                 │
│ {"component":"hook","error":"empty -bugzilla-endpoint, cannot create Bugzilla client","file":"prow/cmd/hook/main.go:145","func":"main.main","level":"fatal","msg":"Error getting Bugzilla client.","severity":"fatal","time":"2020-07-02T07:18:15Z"}                         │
│ stream closed
```
